dot/sync/chain_sync.go

```diff
@@ -53,693 +48,630 @@ func (s chainSyncState) String() string {
-	// a worker that covers the scope of the proposed worker; if true,
-	// ignore the proposed worker
-	hasCurrentWorker(*worker, map[uint64]*worker) bool
-
-	// handleTick handles a timer tick
-	handleTick() ([]*worker, error)
-}
-
 // ChainSync contains the methods used by the high-level service into the `chainSync` module
 type ChainSync interface {
 	start()
 	stop()
 
-	// called upon receiving a BlockAnnounce
-	setBlockAnnounce(from peer.ID, header *types.Header) error
-
 	// called upon receiving a BlockAnnounceHandshake
-	setPeerHead(p peer.ID, hash common.Hash, number uint) error
+	setPeerHead(p peer.ID, hash common.Hash, number uint)
 
 	// syncState returns the current syncing state
 	syncState() chainSyncState
 
 	// getHighestBlock returns the highest block or an error
 	getHighestBlock() (highestBlock uint, err error)
+
+	onImportBlock(announcedBlock) error
+}
+
+type announcedBlock struct {
+	who    peer.ID
+	header *types.Header
 }
 
 type chainSync struct {
-	ctx    context.Context
-	cancel context.CancelFunc
+	stopCh chan struct{}
 
 	blockState BlockState
 	network    Network
 
-	// queue of work created by setting peer heads
-	workQueue chan *peerState
-
-	// workers are put here when they are completed so we can handle their result
-	resultQueue chan *worker
+	workerPool      *syncWorkerPool
+	blockAnnounceCh chan announcedBlock
 
 	// tracks the latest state we know of from our peers,
 	// ie. their best block hash and number
-	sync.RWMutex
-	peerState   map[peer.ID]*peerState
-	ignorePeers map[peer.ID]struct{}
-
-	// current workers that are attempting to obtain blocks
-	workerState *workerState
-
-	// blocks which are ready to be processed are put into this queue
-	// the `chainProcessor` will read from this channel and process the blocks
-	// note: blocks must not be put into this channel unless their parent is known
-	//
-	// there is a case where we request and process "duplicate" blocks, which is where there
-	// are some blocks in this queue, and at the same time, the bootstrap worker errors and dispatches
-	// a new worker with start=(current best head), which results in the blocks in the queue
-	// getting re-requested (as they have not been processed yet)
-	// to fix this, we track the blocks that are in the queue
-	readyBlocks *blockQueue
+	peerViewLock sync.RWMutex
+	peerView     map[peer.ID]*peerView
 
 	// disjoint set of blocks which are known but not ready to be processed
 	// ie. we only know the hash, number, or the parent block is unknown, or the body is unknown
 	// note: the block may have empty fields, as some data about it may be unknown
-	pendingBlocks      DisjointBlockSet
-	pendingBlockDoneCh chan<- struct{}
+	pendingBlocks DisjointBlockSet
 
-	// bootstrap or tip (near-head)
-	state chainSyncState
-
-	// handler is set to either `bootstrapSyncer` or `tipSyncer`, depending on the current
-	// chain sync state
-	handler workHandler
-
-	benchmarker *syncBenchmarker
+	state atomic.Value
```

Comment:
Why was atomic chosen to be used here?

Reply:
The state is a value that can be read among goroutines and can be written only by one of them. So I choose for using atomic since it provides the write and read safely, but it provides the `CompareAndSwap` which only replaces the value inside the atomic if it is different from what is been written, so basically I don't need to hold a lock only for this field and lock/unlock every time we need to check/change the value.